components/datetime/src/format/datetime.rs

```diff
@@ -221,23 +230,62 @@ where
             field.length,
         )?,
         FieldSymbol::DayPeriod(period) => {
-            let symbol = symbols.get_symbol_for_day_period(
-                period,
-                field.length,
-                datetime.datetime().hour().ok_or(Error::MissingInputField)?,
-                arithmetic::is_top_of_hour(
-                    pattern,
-                    datetime.datetime().minute().map(u8::from).unwrap_or(0),
-                    datetime.datetime().second().map(u8::from).unwrap_or(0),
-                ),
-            );
+            let symbol = symbols
+                .expect("Expect symbols to be present")
+                .get_symbol_for_day_period(
+                    period,
+                    field.length,
+                    datetime.datetime().hour().ok_or(Error::MissingInputField)?,
+                    arithmetic::is_top_of_hour(
+                        pattern,
+                        datetime.datetime().minute().map(u8::from).unwrap_or(0),
+                        datetime.datetime().second().map(u8::from).unwrap_or(0),
+                    ),
+                );
             w.write_str(symbol)?
         }
         field @ FieldSymbol::TimeZone(_) => return Err(Error::UnsupportedField(field)),
     };
     Ok(())
 }
 
+// This function determins whether the struct will load symbols data.
+// Keep it in sync with the `write_field` use of symbols.
+pub fn analyze_pattern(pattern: &Pattern, supports_time_zones: bool) -> Result<bool, &Field> {
+    let fields = pattern.items().iter().filter_map(|p| match p {
+        PatternItem::Field(field) => Some(field),
+        _ => None,
+    });
+
+    let mut requires_symbols = false;
+
+    for field in fields {
+        if !requires_symbols {
+            requires_symbols = match field.symbol {
+                FieldSymbol::Month(_) => {
+                    !matches!(field.length, FieldLength::One | FieldLength::TwoDigit)
+                }
+                FieldSymbol::Weekday(_) | FieldSymbol::DayPeriod(_) => true,
+                _ => false,
+            }
```

Comment:
Thought: I have mixed feelings on this.  On the one hand, I think this is the most efficient way to load symbols, and it will get even better when we split each type of symbol into its own key.  On the other hand, I can see the point you've previously made about "chained data loading dependencies", and this sort-of locks us into that model, since we have to load the pattern before we can load the symbols.  That's sort-of where I was doing with my mental model of having two keys, date symbols and time symbols, since we know at compile time whether it is a date-only pattern, a time-only pattern, or a date-time pattern.

Reply:
2021-06-17: This is still an evolving problem. For now, take the 30% win.